timelock-impl/src/main/java/com/palantir/atlasdb/timelock/management/AllNodesDisabledNamespacesUpdater.java

```diff
@@ -0,0 +1,271 @@
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+public class AllNodesDisabledNamespacesUpdater {
+    private static final SafeLogger log = SafeLoggerFactory.get(AllNodesDisabledNamespacesUpdater.class);
+
+    private final AuthHeader authHeader;
+    private final ImmutableList<DisabledNamespacesUpdaterService> remoteUpdaters;
+    private final Map<DisabledNamespacesUpdaterService, CheckedRejectionExecutorService> remoteExecutors;
+    private final TimelockNamespaces localUpdater;
+    private final Supplier<UUID> lockIdSupplier;
+    private final int clusterSize;
+
+    @VisibleForTesting
+    AllNodesDisabledNamespacesUpdater(
+            AuthHeader authHeader,
+            ImmutableList<DisabledNamespacesUpdaterService> remoteUpdaters,
+            Map<DisabledNamespacesUpdaterService, CheckedRejectionExecutorService> remoteExecutors,
+            TimelockNamespaces localUpdater,
+            Supplier<UUID> lockIdSupplier) {
+        this.authHeader = authHeader;
+        this.remoteUpdaters = remoteUpdaters;
+        this.remoteExecutors = remoteExecutors;
+        this.localUpdater = localUpdater;
+        this.lockIdSupplier = lockIdSupplier;
+        this.clusterSize = remoteUpdaters.size() + 1;
+    }
+
+    public static AllNodesDisabledNamespacesUpdater create(
+            AuthHeader authHeader,
+            ImmutableList<DisabledNamespacesUpdaterService> updaters,
+            Map<DisabledNamespacesUpdaterService, CheckedRejectionExecutorService> executors,
+            TimelockNamespaces localUpdater) {
+        return new AllNodesDisabledNamespacesUpdater(authHeader, updaters, executors, localUpdater, UUID::randomUUID);
+    }
+
+    public DisableNamespacesResponse disableOnAllNodes(Set<Namespace> namespaces) {
+        if (anyNodeIsUnreachable()) {
+            return DisableNamespaceResponses.unsuccessfulDueToPingFailure(namespaces);
+        }
+
+        UUID lockId = lockIdSupplier.get();
+        List<SingleNodeUpdateResponse> responses = disableNamespacesOnAllNodes(namespaces, lockId);
+        if (updateWasSuccessfulOnAllNodes(responses)) {
+            return DisableNamespacesResponse.successful(SuccessfulDisableNamespacesResponse.of(lockId));
+        }
+
+        Set<Namespace> consistentlyLockedNamespaces = getConsistentFailures(responses);
+        if (!consistentlyLockedNamespaces.isEmpty()) {
+            return DisableNamespaceResponses.unsuccessfulDueToConsistentlyLockedNamespaces(
+                    consistentlyLockedNamespaces);
+        }
+
+        // If no namespaces were consistently disabled, we should roll back our request,
+        // to avoid leaving ourselves in an inconsistent state.
+        boolean rollbackSuccess = attemptReEnableOnAllNodes(namespaces, lockId, responses.size());
```

Comment:
We only really need to re enable on nodes where we got a success or timed out. If we got a failure, we don't have to bother. Though it's probably fine if it makes code nicer.